File: cpp/day_3.cpp
#include <cmath>
#include <iostream>
#include <fstream>
#include <vector>

using ll = long long;

ll part1(std::vector<std::vector<int>>& lines) {
  ll voltage = 0;
  for (const auto& line: lines) {
    auto lhs = std::max_element(line.begin(), line.end() - 1);
    auto rhs = std::max_element(lhs + 1, line.end());
    voltage += *lhs * 10 + *rhs;
  }
  return voltage;
}

ll part2(std::vector<std::vector<int>>& lines) {
  std::vector<ll> powers;
  for (int i = 0; i <= 11; ++i) {
    powers.push_back(std::pow(10, i));
  }

  ll voltage = 0;
  for (const auto& line: lines) {
    auto left = line.begin(); auto right = line.end() - 11;
    for (int digit = 11; digit >= 0; --digit) {
      auto current = std::max_element(left, right);
      voltage += *current * powers[digit];
      left = current + 1; right += 1;
    }
  }
  return voltage;
}

int main() {
  std::fstream input("../data/input3.txt");
  std::string s;
  std::vector<std::vector<int>> lines;
  
  while (std::getline(input, s)) {
    std::vector<int> line(s.length(), 0);
    for (size_t i = 0; i < s.length(); ++i) {
      line[i] = s[i] - '0';
    }
    lines.push_back(line);
  }

  std::cout << "Part 1: " << part1(lines) << "\n";
  std::cout << "Part 2: " << part2(lines) << "\n";

  return 0;
}

File: rust/src/day_3.rs
use std::fs;

fn main() {
  let input: String = fs::read_to_string("../data/input3.txt")
    .expect("Failed to read file");
  let lines: Vec<Vec<i128>> = parse(&input);
  
  println!("Part 1: {}", part1(&lines));
  println!("Part 2: {}", part2(&lines));
}

fn part1(lines: &[Vec<i128>]) -> i128 {
  let mut voltage: i128 = 0;

  for line in lines {
    let (_, &lhs) = line[..line.len() - 1].iter()
      .enumerate()
      .max_by_key(|&(_, &x)|x)
      .unwrap();
    let &rhs = line.iter().skip_while(|&&x| x != lhs)
      .skip(1).max().unwrap();
    voltage += lhs * 10 + rhs;
  }
  voltage
}

fn part2(lines: &[Vec<i128>]) -> i128 {
  let powers: Vec<i128> = (0..12).map(|i| 10_i128.pow(i)).collect(); 
  let mut voltage = 0;

  for line in lines {
    let mut left = 0;
    let mut right = line.len() - 11;

    for digit in (0..12).rev() {
      let (mut max_i, mut max_v) = (0, &line[left]);
      for (i, v) in line[left..right].iter().enumerate() {
        if v > max_v {
          max_v = v;
          max_i = i;
        }
      }
      voltage += max_v * powers[digit];
      left += max_i + 1;
      right += 1;
    }
  }
  voltage
}

fn parse(input: &str) -> Vec<Vec<i128>> {
  input.lines().map(|line| {
    line.chars().filter_map(|c| c.to_digit(10))
    .map(|d| d as i128).collect()
  }).collect()
}
